[quant][graphmode][fx] Scope support for call_method in QuantizationTracer

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50306 [quant][graphmode][fx] Support preserved_attributes in prepare_fx
* **#50173 [quant][graphmode][fx] Scope support for call_method in QuantizationTracer**

Summary:
Previously we did not set the qconfig for call_method node correctly since it requires us to know
the scope (module path and type of the module whose forward graph contains the node) of the node. This
PR modifies the QuantizationTracer to record the scope information and build a map from call_method
Node to (module_path, module_type), which will be used when we construct qconfig_map

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qconfig_for_call_method
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25818132](https://our.internmc.facebook.com/intern/diff/D25818132)

jerryzh168 committed:
[quant][graphmode][fx] Scope support for call_method in QuantizationTracer

Summary:
Previously we did not set the qconfig for call_method node correctly since it requires us to know
the scope (module path of the module whose forward graph contains the node) of the node. This
PR modifies the QuantizationTracer to record the scope information and build a map from call_method
Node to module path, which will be used when we construct qconfig_map

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qconfig_for_call_method
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Scope support for call_method in QuantizationTracer"


Summary:
Previously we did not set the qconfig for call_method node correctly since it requires us to know
the scope (module path of the module whose forward graph contains the node) of the node. This
PR modifies the QuantizationTracer to record the scope information and build a map from call_method
Node to module path, which will be used when we construct qconfig_map

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qconfig_for_call_method
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25818132](https://our.internmc.facebook.com/intern/diff/D25818132)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Scope support for call_method in QuantizationTracer"


Summary:
Previously we did not set the qconfig for call_method node correctly since it requires us to know
the scope (module path of the module whose forward graph contains the node) of the node. This
PR modifies the QuantizationTracer to record the scope information and build a map from call_method
Node to module path, which will be used when we construct qconfig_map

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qconfig_for_call_method
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25818132](https://our.internmc.facebook.com/intern/diff/D25818132)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Scope support for call_method in QuantizationTracer"


Summary:
Previously we did not set the qconfig for call_method node correctly since it requires us to know
the scope (module path of the module whose forward graph contains the node) of the node. This
PR modifies the QuantizationTracer to record the scope information and build a map from call_method
Node to module path, which will be used when we construct qconfig_map

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qconfig_for_call_method
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25818132](https://our.internmc.facebook.com/intern/diff/D25818132)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Scope support for call_method in QuantizationTracer"


Summary:
Previously we did not set the qconfig for call_method node correctly since it requires us to know
the scope (module path of the module whose forward graph contains the node) of the node. This
PR modifies the QuantizationTracer to record the scope information and build a map from call_method
Node to module path, which will be used when we construct qconfig_map

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qconfig_for_call_method
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25818132](https://our.internmc.facebook.com/intern/diff/D25818132)

[ghstack-poisoned]

jamesr66a reviewed (commented):
LGTM

jerryzh168 committed:
Update on "[quant][graphmode][fx] Scope support for call_method in QuantizationTracer"


Summary:
Previously we did not set the qconfig for call_method node correctly since it requires us to know
the scope (module path and type of the module whose forward graph contains the node) of the node. This
PR modifies the QuantizationTracer to record the scope information and build a map from call_method
Node to (module_path, module_type), which will be used when we construct qconfig_map

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qconfig_for_call_method
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25818132](https://our.internmc.facebook.com/intern/diff/D25818132)

[ghstack-poisoned]

vkuzo commented on test/quantization/test_quantize_fx.py:
```diff
@@ -915,6 +914,49 @@ def forward(self, x):
+
+
+        m = M().eval()
+        qconfig_dict = {"": default_qconfig, "module_name": [("sub", None)]}
```
should we test the reverse as well?

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -346,50 +344,47 @@ def _generate_qconfig_map(
-            qconfig_dict: Any) -> None:
-        global_qconfig = qconfig_dict.get('', None)
+            qconfig_dict: Any,
+            node_name_to_scope: Dict[str, Tuple[str, Any]]) -> None:
```
any chance we can clarify `Any` further to the real type?  Motivation: if someone other than the code author is debugging this, as written there is no way to deduce the type other than looking up this PR and tracing what's happening.  If the type was specified, the code reader has a lower burden.

if it's challening to define the type for some reason, can we add a comment with an example value

jerryzh168 replied:
sure

vkuzo commented on torch/quantization/quantize_fx.py:
```diff
@@ -41,20 +42,99 @@ def _fuse_fx(
             type(m) in self.skipped_module_classes or \
             isinstance(m, _FusedModule)
 
+    def call_module(self, m: torch.nn.Module, forward: Callable[..., Any], args : Tuple[Any, ...], kwargs : Dict[str, Any]) -> Any:
+        module_qualified_name = self.path_of_module(m)
+        if not self.is_leaf_module(m, module_qualified_name):
+            def scoped_forward(_args, _kwargs):
+                # Creating scope with information of current module
+                # scope will be restored automatically upon exit
+                with ScopeContextManager(self.scope, m, module_qualified_name):
+                    return forward(*_args, **_kwargs)
+            return scoped_forward(args, kwargs)
+        return self.create_proxy("call_module", module_qualified_name, args, kwargs)
+
+    def create_node(self, kind : str, target : Target,
+                    args : Tuple[Argument, ...], kwargs : Dict[str, Argument], name : Optional[str] = None,
+                    type_expr : Optional[Any] = None) -> Node:
+        node = super().create_node(kind, target, args, kwargs, name, type_expr)
+        if kind == "call_method":
+            self.node_name_to_scope[node.name] = (self.scope.module_path, self.scope.module_type)
+        return node
```
do we need any test coverage that these still satisfy the contract of the parent, in case the parent behavior changes?  I'm not saying they do (I'm the wrong person to ask), probably more a question for the FX team.

jerryzh168 replied:
yeah James has reviewed the PR. I think there might be some concerns if the code for "call_module" changes, I can check with him again.

vkuzo commented on torch/quantization/quantize_fx.py:
```diff
@@ -41,20 +42,66 @@ def _fuse_fx(
+        self.module_path = module_path
+        self.module_type = module_type
+
+class ScopeContextManager(object):
```
outdated

jerryzh168 commented on torch/quantization/quantize_fx.py:
```diff
@@ -41,20 +42,99 @@ def _fuse_fx(
             isinstance(m, _FusedModule)
 
+    def call_module(self, m: torch.nn.Module, forward: Callable[..., Any], args : Tuple[Any, ...], kwargs : Dict[str, Any]) -> Any:
+        module_qualified_name = self.path_of_module(m)
```
@jamesr66a this is copied from default `call_module` code, would it be a problem if the default code changes? do you think it's better to provide an API in `call_module` to allow this kind of extension?

jamesr66a replied:
I think the future-proofed way to do this is to instantiate the guard and call `super().call_module`

jerryzh168 replied:
OK, sg

jerryzh168 committed:
Update on "[quant][graphmode][fx] Scope support for call_method in QuantizationTracer"


Summary:
Previously we did not set the qconfig for call_method node correctly since it requires us to know
the scope (module path and type of the module whose forward graph contains the node) of the node. This
PR modifies the QuantizationTracer to record the scope information and build a map from call_method
Node to (module_path, module_type), which will be used when we construct qconfig_map

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qconfig_for_call_method
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25818132](https://our.internmc.facebook.com/intern/diff/D25818132)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Scope support for call_method in QuantizationTracer"


Summary:
Previously we did not set the qconfig for call_method node correctly since it requires us to know
the scope (module path and type of the module whose forward graph contains the node) of the node. This
PR modifies the QuantizationTracer to record the scope information and build a map from call_method
Node to (module_path, module_type), which will be used when we construct qconfig_map

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qconfig_for_call_method
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25818132](https://our.internmc.facebook.com/intern/diff/D25818132)

[ghstack-poisoned]
